File: android/leokit/src/main/kotlin/io/github/leoho0722/leokit/components/LKPickerField.kt
package io.github.leoho0722.leokit.components

import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import io.github.leoho0722.leokit.LKShapes
import io.github.leoho0722.leokit.LKTheme
import io.github.leoho0722.leokit.tokens.LKSize
import io.github.leoho0722.leokit.tokens.LKSpacing

/**
 * 一個看起來像輸入框、按下去會打開原生選擇器的欄位。
 *
 * 這個元件只是「欄位」。它不持有值，也不自己彈出任何東西：
 * 選擇面請用 `ExposedDropdownMenuBox`、`DatePickerDialog` 或 [LKCombobox]。
 * 日期與時間一律用平台原生選擇器：時區、農曆、週起始日與語系格式系統都處理好了。
 * [value] 傳已經格式化好的字串，格式化規則屬於呼叫端的業務邏輯。
 * 選項只有兩三個且需要同時看見時，用 [LKRadio] 會比打開選擇面更快。
 *
 * @param label 欄位上方的標籤
 * @param onClick 按下欄位時要做的事，通常是打開原生選擇器
 * @param modifier 由呼叫端套在整個欄位上的修飾子
 * @param value 已經格式化好的顯示字串；還沒選時傳 null
 * @param placeholder 還沒選時顯示的提示文字
 * @param kind 這個欄位會打開哪一種原生選擇器，決定前面畫什麼圖示
 * @param help 欄位下方的說明；[errorText] 有值時會被蓋掉
 * @param errorText 錯誤訊息；有值時整格轉為錯誤樣式
 * @param enabled 這個欄位目前是否可以按
 */
@Composable
public fun LKPickerField(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    value: String? = null,
    placeholder: String = "請選擇",
    kind: LKPickerFieldKind = LKPickerFieldKind.Options,
    help: String? = null,
    errorText: String? = null,
    enabled: Boolean = true,
) {
    val colors = LKTheme.colors
    val typography = LKTheme.typography
    val isError = errorText != null
    val leading = when (kind) {
        LKPickerFieldKind.Options -> null
        LKPickerFieldKind.Date -> LKIcons.Calendar
        LKPickerFieldKind.Time -> LKIcons.Clock
    }

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(LKSpacing.spacing4),
    ) {
        Text(text = label, style = typography.subhead, color = colors.textSecondary)
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .defaultMinSize(minHeight = LKSize.controlHMd)
                .border(
                    width = if (isError) LKSize.borderWStrong else LKSize.hairline,
                    color = if (isError) colors.danger else colors.borderControl,
                    shape = LKShapes.md,
                )
                .clickable(enabled = enabled, role = Role.Button, onClick = onClick)
                // label 是這個 Row 的 sibling，不寫進來的話 TalkBack 聚焦到這顆按鈕時
                // 只唸得到值，畫面上有兩個以上的欄位就分不出誰是誰
                .semantics(mergeDescendants = true) {
                    contentDescription = "$label，${value ?: placeholder}"
                    if (errorText != null) {
                        error(errorText)
                    }
                }
                .padding(horizontal = LKSpacing.spacing12),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(LKSpacing.spacing8),
        ) {
            if (leading != null) {
                Icon(
                    imageVector = leading,
                    contentDescription = null,
                    tint = colors.textTertiary,
                    modifier = Modifier.size(LKSize.iconMd),
                )
            }
            Text(
                text = value ?: placeholder,
                style = typography.body,
                color = if (value == null) colors.textTertiary else colors.textPrimary,
                modifier = Modifier.weight(1f),
            )
            Icon(
                imageVector = LKIcons.ChevronDown,
                contentDescription = null,
                tint = colors.textTertiary,
                modifier = Modifier.size(LKSize.iconSm),
            )
        }
        when {
            errorText != null -> Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(LKSpacing.spacing4),
            ) {
                Icon(
                    imageVector = LKIcons.Alert,
                    contentDescription = null,
                    tint = colors.dangerInk,
                    modifier = Modifier.size(LKSize.iconSm),
                )
                Text(text = errorText, style = typography.footnote, color = colors.dangerInk)
            }

            help != null ->
                Text(text = help, style = typography.footnote, color = colors.textSecondary)
        }
    }
}

/** 這個欄位會打開哪一種原生選擇器，決定它前面畫什麼圖示。 */
public enum class LKPickerFieldKind {
    /** 從一份清單裡選一個。 */
    Options,

    /** 選一個日期。 */
    Date,

    /** 選一個時間。 */
    Time,
}

File: ios/Sources/LeoKit/Components/LKPickerField.swift
//
//  LKPickerField.swift
//  LeoKit
//
//  Created by Leo Ho on 2026/09/20.
//

import SwiftUI

/// 一個看起來像輸入框、按下去會打開原生選擇器的欄位
///
/// - Note: 這個元件只是「欄位」。它不持有值，也不自己彈出任何東西 ——
///   選擇面一律交給系統的 `Picker`、`DatePicker` 或 `LKCombobox`
/// - Note: 日期與時間一律用平台原生選擇器，時區、農曆、週起始日與語系格式系統都處理好了
/// - Note: `value` 傳已經格式化好的字串；格式化規則屬於呼叫端的業務邏輯
/// - Note: 選項只有兩三個且需要同時看見時，用 `LKRadio` 會比打開選擇面更快
public struct LKPickerField: View {

    // MARK: - Properties

    /// 按下欄位時要做的事，通常是打開原生選擇器
    private let action: () -> Void

    /// 錯誤訊息；有值時整格轉為錯誤樣式
    private let error: String?

    /// 欄位下方的說明；`error` 有值時會被蓋掉
    private let help: String?

    /// 這個欄位會打開哪一種原生選擇器
    private let kind: LKPickerFieldKind

    /// 還沒選時顯示的提示文字
    private let placeholder: String

    /// 欄位上方的標籤
    private let title: String

    /// 已經格式化好的顯示字串；還沒選時是 nil
    private let value: String?

    // MARK: - Init

    /// 建立一個會打開原生選擇器的欄位
    ///
    /// - Parameters:
    ///   - title: 欄位上方的標籤
    ///   - value: 已經格式化好的顯示字串
    ///   - placeholder: 還沒選時顯示的提示文字
    ///   - kind: 這個欄位會打開哪一種原生選擇器
    ///   - help: 欄位下方的說明
    ///   - error: 錯誤訊息
    ///   - action: 按下欄位時要做的事
    public init(
        _ title: String,
        value: String? = nil,
        placeholder: String = "請選擇",
        kind: LKPickerFieldKind = .options,
        help: String? = nil,
        error: String? = nil,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.value = value
        self.placeholder = placeholder
        self.kind = kind
        self.help = help
        self.error = error
        self.action = action
    }

    // MARK: - Body

    /// 骨架：標籤、可按的欄位、下方的說明或錯誤
    public var body: some View {
        VStack(alignment: .leading, spacing: LKSpacing.spacing4) {
            titleText

            field

            footer
        }
        .accessibilityElement(children: .contain)
    }
}

// MARK: - Private Views

private extension LKPickerField {

    /// 欄位上方的標籤
    ///
    /// - Note: 對 VoiceOver 隱藏 —— 它已經是欄位本身的名稱了，再唸一次是重複
    var titleText: some View {
        Text(title)
            .font(LKFont.subhead)
            .foregroundStyle(LKColor.textSecondary)
            .accessibilityHidden(true)
    }

    /// 可以按的欄位本體：圖示、目前的值、尾端的箭頭
    var field: some View {
        Button(action: action) {
            HStack(spacing: LKSpacing.spacing8) {
                leadingIcon

                Text(value ?? placeholder)
                    .font(LKFont.body)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .foregroundStyle(value == nil ? LKColor.textTertiary : LKColor.textPrimary)

                Image(systemName: "chevron.down")
                    .font(.system(size: LKSize.iconSm, weight: .semibold))
                    .foregroundStyle(LKColor.textTertiary)
            }
            .padding(.horizontal, LKSpacing.spacing12)
            .frame(minHeight: LKSize.controlHMd)
            .background(LKColor.bgSurface, in: LKShape.md)
            .overlay {
                LKShape.md.stroke(borderColor, lineWidth: borderWidth)
            }
            .contentShape(.rect)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(title)
        .accessibilityValue(value ?? placeholder)
        .accessibilityHint(error ?? help ?? "")
    }

    /// 欄位最前面的圖示，清單型不畫
    @ViewBuilder
    var leadingIcon: some View {
        if let systemImage = kind.systemImage {
            Image(systemName: systemImage)
                .font(.system(size: LKSize.iconMd))
                .foregroundStyle(LKColor.textTertiary)
        }
    }

    /// 欄位下方的一行字：有錯誤就顯示錯誤，否則顯示說明
    @ViewBuilder
    var footer: some View {
        if let error {
            Label {
                Text(error)
                    .font(LKFont.footnote)
            } icon: {
                Image(systemName: "exclamationmark.triangle")
            }
            .font(LKFont.footnote)
            .foregroundStyle(LKColor.dangerInk)
        } else if let help {
            Text(help)
                .font(LKFont.footnote)
                .foregroundStyle(LKColor.textSecondary)
        }
    }
}

// MARK: - Computed Properties

private extension LKPickerField {

    /// 外框的顏色
    var borderColor: Color {
        if error == nil {
            LKColor.borderControl
        } else {
            LKColor.danger
        }
    }

    /// 外框的粗細，有錯誤時加粗
    var borderWidth: CGFloat {
        if error == nil {
            LKSize.hairline
        } else {
            LKSize.borderWStrong
        }
    }
}

// MARK: - Preview

#Preview("三種欄位") {
    VStack(spacing: LKSpacing.spacing16) {
        LKPickerField("扣款週期", value: "每月", help: "改成年繳通常會便宜一些") { }

        LKPickerField("下次扣款日", value: "2026/10/03", kind: .date) { }

        LKPickerField("提醒時間", kind: .time, error: "請選一個提醒時間") { }
    }
    .padding(LKSpacing.spacing16)
    .background(LKColor.bgCanvas)
}
